### ISSUE TITLE
WSL not detected

### ISSUE THREAD

**Issue by daldr-ntml** [microsoft/vscode-cpptools#2644]
**Type: LanguageService**

- OS and Version: Windows 10 Pro 1803
- VS Code Version: 1.28.0
- C/C++ Extension Version: 0.19.0

I alluded to this problem in issue #2596 but this is a distinct problem so I'm opening a new issue for it here.

My c_cpp_properties.json contains:

            "intelliSenseMode": "clang-x64",
            "compilerPath": "/usr/bin/gcc",
            "cStandard": "c11",
            "cppStandard": "c++17"

I have Ubuntu 18.04 WSL installed. The registry contains the WSL key:

![image](https://user-images.githubusercontent.com/7724321/46789264-f27b6780-cd33-11e8-867b-bb9b1e4ba06f.png)

When I start VS Code the extension reports 'WSL not detected':

    initialized
    workspace/didChangeConfiguration
    IntelliSense Engine = Default.
    The extension will use the Tag Parser for IntelliSense when #includes don't resolve.
    Autocomplete is enabled.
    Error squiggles are enabled.
    File exclude: **/.git
    File exclude: **/.svn
    File exclude: **/.hg
    File exclude: **/CVS
    File exclude: **/.DS_Store
    File exclude: **/*.o
    File exclude: **/.BROWSE2*
    File exclude: **/*.log
    File exclude: **/*.tlog
    File exclude: **/*.d
    File exclude: **/.vscode
    Search exclude: **/node_modules
    Search exclude: **/bower_components
    Search exclude: **/*.vcxproj
    Search exclude: **/*.project
    Search exclude: **/prototypes
    Search exclude: **/.vscode
    cpptools/queryCompilerDefaults
    Attempting to get defaults from compiler found on the machine: 'C:/VS2017BuildTools/VC/Tools/MSVC/14.15.26726/bin/Hostx64/x64/cl.exe'
    cpptools/didChangeFolderSettings
    Attempting to get defaults from compiler in "compilerPath" property: '/usr/bin/gcc'
    Failed to query default include paths and defines for /usr/bin/gcc. WSL not detected
    Attempting to get defaults from compiler found on the machine: /usr/bin/gcc
    Code browsing service initialized
    Attempting to get defaults from compiler in "compilerPath" property: '/usr/bin/gcc'
    Failed to query default include paths and defines for /usr/bin/gcc. WSL not detected
    Attempting to get defaults from compiler found on the machine: /usr/bin/gcc
      Folder: C:/VS2017BUILDTOOLS/VC/TOOLS/MSVC/14.15.26726/INCLUDE/* will be indexed
      Folder: C:/PROGRAM FILES (X86)/WINDOWS KITS/10/INCLUDE/10.0.17134.0/UM/ will be indexed
      Folder: C:/PROGRAM FILES (X86)/WINDOWS KITS/10/INCLUDE/10.0.17134.0/UCRT/ will be indexed
      Folder: C:/PROGRAM FILES (X86)/WINDOWS KITS/10/INCLUDE/10.0.17134.0/SHARED/ will be indexed
      Folder: C:/PROGRAM FILES (X86)/WINDOWS KITS/10/INCLUDE/10.0.17134.0/WINRT/ will be indexed
      Folder: C:/PROGRAM FILES (X86)/WINDOWS KITS/10/INCLUDE/10.0.17134.0/CPPWINRT/ will be indexed
      Folder: C:/LEDA_SDK/SYSROOTS/HASWELL-64-WRS-LINUX/USR/INCLUDE/ will be indexed
      Folder: C:/SVNPROJ/LEDA/TRUNK/SOFTWARE/ will be indexed
    Discovering files...
      Processing folder (non-recursive): C:/VS2017BUILDTOOLS/VC/TOOLS/MSVC/14.15.26726/INCLUDE
      Processing folder (recursive): C:/PROGRAM FILES (X86)/WINDOWS KITS/10/INCLUDE/10.0.17134.0/UM/
      Processing folder (recursive): C:/PROGRAM FILES (X86)/WINDOWS KITS/10/INCLUDE/10.0.17134.0/UCRT/
      Processing folder (recursive): C:/PROGRAM FILES (X86)/WINDOWS KITS/10/INCLUDE/10.0.17134.0/SHARED/
      Processing folder (recursive): C:/PROGRAM FILES (X86)/WINDOWS KITS/10/INCLUDE/10.0.17134.0/WINRT/
      Processing folder (recursive): C:/PROGRAM FILES (X86)/WINDOWS KITS/10/INCLUDE/10.0.17134.0/CPPWINRT/
      Processing folder (recursive): C:/LEDA_SDK/SYSROOTS/HASWELL-64-WRS-LINUX/USR/INCLUDE/
      Processing folder (recursive): C:/SVNPROJ/LEDA/TRUNK/SOFTWARE/
      Discovering files: 15670 file(s) processed
      0 file(s) removed from database
    Done discovering files.
    Populate include completion cache.
    cpptools/fileCreated
    cpptools/fileCreated
    Parsing remaining files...
      tag parsing file: C:\SVNPROJ\LEDA\TRUNK\SOFTWARE\SOURCE\LEDAAP\INCLUDES\MSGPROTOCOL.H
      tag parsing file: C:\SVNPROJ\LEDA\TRUNK\SOFTWARE\SOURCE\LEDAAP\INCLUDES\BFEMULATIONPCIF.H
      tag parsing file: C:\SVNPROJ\LEDA\TRUNK\SOFTWARE\SOURCE\LEDAAP\MAIN.CPP
      tag parsing file: C:\SVNPROJ\LEDA\TRUNK\SOFTWARE\BEAMFORMING\BF_PRECODER_SW\BF_PRECODER\HOST\INC\BF_PRECODER.H
      tag parsing file: C:\SVNPROJ\LEDA\TRUNK\SOFTWARE\BEAMFORMING\COMMON\INC\AOCLUTILS\AOCL_UTILS.H
      tag parsing file: C:\SVNPROJ\LEDA\TRUNK\SOFTWARE\SOURCE\LEDACOMMON\INCLUDES\BUFFERMANAGER.H
      tag parsing file: C:\SVNPROJ\LEDA\TRUNK\SOFTWARE\SOURCE\LEDAAP\BF\INCLUDES\WEIGHTS.H
      tag parsing file: C:\SVNPROJ\LEDA\TRUNK\SOFTWARE\DRIVER\PCIE\LEDAPCI.H
      tag parsing file: C:\SVNPROJ\LEDA\TRUNK\SOFTWARE\DRIVER\PCIE\INCLUDE\BWPCI_DEF.H
      tag parsing file: C:\SVNPROJ\LEDA\TRUNK\SOFTWARE\SOURCE\LEDACOMMON\INCLUDES\FILEIO.H
      tag parsing file: C:\SVNPROJ\LEDA\TRUNK\SOFTWARE\SOURCE\LEDACOMMON\INCLUDES\LINKEDDOUBLYLIST.H
      tag parsing file: C:\SVNPROJ\LEDA\TRUNK\SOFTWARE\SOURCE\LEDACOMMON\INCLUDES\TIMESYNC.H
      tag parsing file: C:\SVNPROJ\LEDA\TRUNK\SOFTWARE\SOURCE\LEDACOMMON\INCLUDES\TIMING.H
      Parsing: 13 files(s) processed
    Done parsing remaining files.
    Database safe to open

So the Windows compiler is used, which I do not want.  Why is WSL not detected?

**Comment by sean-mcmanus** [comment]
This is very strange. I think we might need to add more logging and error checking.

**Comment by bobbrow** [comment]
What is the value of "DefaultDistribution" in the "Lxss" Key?

**Comment by daldr-ntml** [comment]
The value of "DefaultDistribution" in the "Lxss" Key is {7005b00a-3fc2-4f43-baeb-debc85d6a86e}

![image](https://user-images.githubusercontent.com/7724321/46936099-22d54580-d055-11e8-8ade-e911f7ab9adb.png)

**Comment by bobbrow** [comment]
It looks like we can't find your distro because there's no key for it in the registry.  You should be able to use the `wslconfig.exe` tool in a command shell to correct this.

`wslconfig /l` - list the known distros
`wslconfig /s <distro name>` - set the default distro

**Comment by bobbrow** [comment]
You might also consider reporting this registry error to https://github.com/Microsoft/WSL.

**Comment by sean-mcmanus** [comment]
@DavidA2014 Do you have any idea how this non-existent DefaultDistribution got set? Did you remove a distro? If this issue is more widespread maybe we could check to see if the DefaultDistribution GUID actually exists and if not, fall back to an existing one.

**Comment by daldr-ntml** [comment]
Thanks for your help again. I set the default distribution as Bob suggested:

    C:\>wslconfig /l
    Windows Subsystem for Linux Distributions:
    Ubuntu-18.04 (Default)
    
    C:\>wslconfig /s Ubuntu-18.04
    
    C:\>

and now cpptools detects WSL, so the problem is solved.

I don't know 'how this non-existent DefaultDistribution got set'.  I did have trouble installing the distro from the Microsoft store because local IT policy initially prevented access to the store. That was changed but perhaps something strange did happen during the installation.  I could report it to the WSL group but I haven't got specific steps to reproduce it.

Anyway, I now have a working gcc compiler that doesn't define _WIN32!  Thanks very much for your help.
